CS3219-AY2223S1/cs3219-project-ay2223s1-g12
# Implement pending match model and ORM layer

Fixes #13, #15 

Implementations:
- Create database connection with nodejs
- Implement the pending match model 
- Implement the ORM layer. This layer will serve as an 'intermediary' between the controller and the database. 

Manual testing:
1. `cd matching-service`
2. `npm run dev` to start the server
3. Check that server is running at `http://localhost:8001` You can verify this by:
    3.1 access port `8001` locally  or
    3.2 send a `GET` request through *Postman Desktop Agent*

    At this step, 'Hello Wold from matching-service' should be visible
4.  From *Postman Desktop Agent* send a `POST` request to `http://localhost:8001/pendingMatches` containing a pending match, e.g.:
```json
{
   "username":  "hello",
   "difficulty":  "easy"
}
```
This should be successful, with HTTP status code `200` with the object in the return body.

(Tested using username 'punpun')

![94E62CBB-CAED-4E1C-BC87-399DB07E9E9C](https://user-images.githubusercontent.com/60144099/188284692-f791026c-1406-4a74-b85a-4141885b8771.jpeg)

5.  Sending the exact pending match with the same username again i.e. *hello* in this case, should throw an error (this is because our database should not have a duplicate username, which will affect matching, currently I use username as the primary key)

(Tried using username 'punpun' again)

![18CE58C0-BD82-4733-A431-AE3C698A7FD6](https://user-images.githubusercontent.com/60144099/188284695-64cde9fa-77b2-402a-8954-412768d9a818.jpeg)
---------
Things that can be done in another commit:
- Decide whether username should really be a primary key, or should we use id as primary key instead (i.e. add another column in the database)
- Add checks for the difficulty levels (i.e. should only be easy, medium, hard). Not sure if this can be done at the frontend side instead, or maybe both frontend and database.
- Add more methods in the ORM layer for use in the future (e.g. retrieve all users in the database, update difficulty level when the user decide to change difficulty)
----------
(RESOLVED) Question:
Normally package-lock.json should not be committed right? It is originally included from the GitHub classroom initial commit. I removed it here. But to verify if this is ok.
